sw/device/lib/base/memory.h

```diff
@@ -78,6 +79,9 @@ inline ptrdiff_t misalignment32_of(uintptr_t addr) {
  */
 OT_WARN_UNUSED_RESULT
 inline uint32_t read_32(const void *ptr) {
+  if (ptr == NULL) {
+    OT_UNREACHABLE();
+  }
```

Comment:
can we remove this check? asking because we use this function in silicon_creator code where we omit such checks.

Reply:
We can, but it lets clang-analyzer generate some non-fatal warnings:


```
./sw/device/lib/base/memory.h:91:3: warning: Null pointer passed as 2nd argument to memory copy function [clang-analyzer-unix.cstring.NullArg]
  __builtin_memcpy(&val, ptr, sizeof(uint32_t));
  ^
/home/dan_zerorisc_com/.cache/bazel/_bazel_dan/e8b9a1cec6d6126ab719828a176ad974/sandbox/linux-sandbox/35537/execroot/lowrisc_opentitan/sw/device/lib/base/memory.c:150:10: note: Assuming 'end' is <= 'tail_offset'
  for (; end > tail_offset; --end) {
         ^~~~~~~~~~~~~~~~~
/home/dan_zerorisc_com/.cache/bazel/_bazel_dan/e8b9a1cec6d6126ab719828a176ad974/sandbox/linux-sandbox/35537/execroot/lowrisc_opentitan/sw/device/lib/base/memory.c:150:3: note: Loop condition is false. Execution continues on line 158
  for (; end > tail_offset; --end) {
  ^
/home/dan_zerorisc_com/.cache/bazel/_bazel_dan/e8b9a1cec6d6126ab719828a176ad974/sandbox/linux-sandbox/35537/execroot/lowrisc_opentitan/sw/device/lib/base/memory.c:158:10: note: Assuming 'end' is > 'body_offset'
  for (; end > body_offset; end -= sizeof(uint32_t)) {
         ^~~~~~~~~~~~~~~~~
/home/dan_zerorisc_com/.cache/bazel/_bazel_dan/e8b9a1cec6d6126ab719828a176ad974/sandbox/linux-sandbox/35537/execroot/lowrisc_opentitan/sw/device/lib/base/memory.c:158:3: note: Loop condition is true.  Entering loop body
  for (; end > body_offset; end -= sizeof(uint32_t)) {
  ^
/home/dan_zerorisc_com/.cache/bazel/_bazel_dan/e8b9a1cec6d6126ab719828a176ad974/sandbox/linux-sandbox/35537/execroot/lowrisc_opentitan/sw/device/lib/base/memory.c:161:27: note: Calling 'read_32'
    uint32_t word_right = read_32(&rhs8[i]);
                          ^~~~~~~~~~~~~~~~~
./sw/device/lib/base/memory.h:89:3: note: Null pointer value stored to 'ptr'
  ptr = __builtin_assume_aligned(ptr, alignof(uint32_t));
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sw/device/lib/base/memory.h:91:3: note: Null pointer passed as 2nd argument to memory copy function
  __builtin_memcpy(&val, ptr, sizeof(uint32_t));
  ^                      ~~~
```

Reply:
@alphan WDYT about moving `  if (ptr == NULL) { OT_UNREACHABLE(); }` to a handful of callsites in memory.c?